### DISCUSSION TITLE
Cluster plugin adding 100+ms latency

### DISCUSSION THREAD

**Discussion by anup-krai** [Kong/kong#10676] [Help]
I am using a custom plugin which will run at cluster level and will add extra header if not present. Below is the plugin code -
```
function SetRegionHeader:rewrite(conf)
    local region = kong.request.get_header(conf.header_name)
    if not region or region == "" then
      region = os.getenv("KONG_REGION")
      if region then
        kong.service.request.set_header(conf.header_name, region)
      end
    end
  end
```
If I enable this plugin at cluster level then I see Kong proxy latency is getting increased by 100+ms. 
I am using Kong 2.7.1 with KIC 2.2.1 with Postgres DB.
Anything in this code which is increasing this latency?

**Comment by ghost** [comment]
`os.getenv` is a system command line call and a blocking operation. You can try replacing “region” with a constant string to see if the delay is related to it.

**Reply by anup-krai** [reply]
I hard coded the value of “region” but still I see the same latency.

**Reply by ghost** [reply]
You can use binary search to gradually locate the code that causes the delay.

**Reply by anup-krai** [reply]
Can you please give an example on how we can do this?

**Reply by samugi** [reply]
@anup-krai I think @ghost meant to recommend investigating your custom plugin's code by running it empty at first (e.g. comment all the logic), then gradually uncomment lines to narrow down the source of the latency
